src/data/program_configuration.js

```diff
@@ -129,31 +140,49 @@ class SourceExpressionBinder<T> implements Binder<T> {
         return [];
     }
 
-    populatePaintArray(length: number, feature: Feature) {
+    populatePaintArray(newLength: number, feature: Feature) {
         const paintArray = this.paintVertexArray;
 
         const start = paintArray.length;
-        paintArray.reserve(length);
+        paintArray.reserve(newLength);
 
         const value = this.expression.evaluate({zoom: 0}, feature);
 
         if (this.type === 'color') {
             const color = packColor(value);
-            for (let i = start; i < length; i++) {
+            for (let i = start; i < newLength; i++) {
                 paintArray.emplaceBack(color[0], color[1]);
             }
         } else {
-            for (let i = start; i < length; i++) {
+            for (let i = start; i < newLength; i++) {
                 paintArray.emplaceBack(value);
             }
 
             this.statistics.max = Math.max(this.statistics.max, value);
         }
     }
 
+    updatePaintArray(start: number, end: number, feature: Feature) {
+        const paintArray = this.paintVertexArray;
+        const value = this.expression.evaluate({zoom: 0}, feature);
+
+        if (this.type === 'color') {
+            const color = packColor(value);
+            for (let i = start; i < end; i++) {
+                paintArray.emplace(i, color[0], color[1]);
+            }
+        } else {
+            for (let i = start; i < end; i++) {
+                paintArray.emplace(i, value);
+            }
+
+            this.statistics.max = Math.max(this.statistics.max, value);
```

Comment:
This doesn't handle the case where the maximum value is replaced with a smaller value and the max should decrease. I think this is fine though, because having a bigger max won't cause bugs the way we use it and maintaining the real max value would be more complicated and expensive. This should be fine and safe the way it is, right?

Reply:
Keeping track of the `max` value correctly may not be worth it. When a property value is reduced, the update logic would need to filter the previously populated values to check if there is a new lower `max`. 

Leaving this as is won't return incorrect results, but it will increase the number of features that are matched in the initial query (with `queryPadding`) and then have to be filtered out. 

If there are use cases that significantly affect the performance of `queryRenderedFeatures`, this can be revisited.